### PR TITLE
Phase 3, release 1

### PR DESCRIPTION
**Pull request by calvogenerico**
# Phase 3, release 1:

- Start regular upgrades
    - Added logic to initiate regular upgrades. Now, when an upgrade is approved in the ZkProtocolGovernor in L2, it can be easily started within the ProtocolUpgradeHandler from the webapp.
- New repo structure
    - Repository is now organized into apps and packages, facilitating code sharing between apps.
    - Added Turborepo for better code sharing, faster build times, and improved developer experience.
- Added end-to-end (E2E) tests for webapp.
    - Implemented Playwright + Dappwright for testing, with custom fixes.
    - Test suite covers every flow in the app.
    - CI executes tests on every pull request.
- Performance improvements
    - Implemented changes throughout the app to eliminate redundant RPC calls and slow operations.
    - Optimized UI rendering to reduce blocking whenever possible.
- Enhanced signature verification strategy
    - Improved code consistency: The logic for building the signature pre-image is the same in both backend and frontend, reducing the potential for future bugs.
    - Ensured data freshness: All data used to validate signatures is now gathered just before validation. This guarantees accurate signature validation throughout the voting period, even if changes occur.
- Refactors and improvements
    - Eliminated duplicated code and logic.
    - Streamlined dependency list, reducing overall dependencies.
    - Standardized dependency versions for both CLI and webapp.
    - Significantly improved development environment.
    - Optimized CI process, executing long tasks only when necessary.

### CODE REVIEW

**Comment by tomimor** [comment]
LGTM ⭐

File: apps/web/app/.server/service/signatures.ts [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+import { guardianMembers } from "./ethereum-l1/contracts/guardians";
+import { securityCouncilMembers } from "./ethereum-l1/contracts/security-council";
+import { zkFoundationAddress } from "./ethereum-l1/contracts/emergency-upgrade-board";
+import {
+  emergencyUpgradeBoardAddress,
+  guardiansAddress,
+  securityCouncilAddress,
+} from "./ethereum-l1/contracts/protocol-upgrade-handler";
+
+async function shouldMarkProposalAsReady(allSignatures: BasicSignature[]): Promise<boolean> {
```
Comment (anton-matterlabs) [inline apps/web/app/.server/service/signatures.ts:43]: This seems to be emergency upgrade specific. Let's reflect it in the function name.

**Review by anton-matterlabs — COMMENTED** [review]
I did my best, but I am not a Typescript expert and the diff is huge, so this review is rather superficial.

File: .github/workflows/docker-reusable.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Build Docker image
+        id: build
+        uses: docker/build-push-action@v6
+        with:
+          context: .
+          push: ${{ github.ref == 'refs/heads/main' }}
+          tags: |
+            ghcr.io/${{ github.repository }}:latest
+            ghcr.io/${{ github.repository }}:${{ env.SHORT_SHA }}-${{ env.UNIX_TIME }}
```
[the remark this answers is not in the document]
Comment (aon) [reply .github/workflows/docker-reusable.yml]: Indeed, it's not really necessary. Removed in 1442a930a7ba7b279816ee9a866075295c71ff59

File: .github/workflows/test-e2e-reusable.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      - run: cd ../contracts && pnpm build
+      - run: pnpm test:web:env:up
+      - run: xvfb-run pnpm playwright test
+      - uses: actions/upload-artifact@v4
+        if: ${{ !cancelled() }}
+        with:
+          name: playwright-report
+          path: packages/e2e/playwright-report/
+          retention-days: 30
+      - if: always()
```
[the remark this answers is not in the document]
Comment (aon) [reply .github/workflows/test-e2e-reusable.yml:68]: Yes, both. We want to always print logs, if it fails or not. In the same way, we applied it to the app shutdown.

File: package.json [hunk 4]
```diff
@@ -1,35 +1,28 @@
...
-    "pnpm": ">=8"
-  },
-  "files": [
-    "dist",
-    "cli/bin",
-    "*.d.ts",
-    "*.mjs"
-  ],
+  "author": "moonsong-labs",
+  "license": "GPL-3.0-only",
```
[the remark this answers is not in the document]
Comment (aon) [reply package.json:7]: What license should we have here? Or should we leave it empty?

[the remark this answers is not in the document]
Comment (anton-matterlabs) [reply package.json:7]: Our recommendation is Apache 2.0 or dual MIT/Apache 2.0, same as e.g. Rust standard: https://www.rust-lang.org/policies/licenses

File: apps/web/app/.server/service/proposals.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+      l1ProposalId: null;
+    }
+);
+
+async function extractProposalData(txHash: Hex, l2ProposalId: Hex): Promise<ProposalDataResponse> {
+  console.log("txHash", txHash);
+  const receipt = await l2Rpc.getTransactionReceipt({ hash: txHash });
+  const logProof = await fetchLogProof(txHash, 0);
+
+  const l1MessageEventId = "0x3a36e47291f4201faf137fab081d92295bce2d53be2c6ca68ba82c7faa9ce241";
```
[the remark this answers is not in the document]
Comment (aon) [reply apps/web/app/.server/service/proposals.ts]: Fixed in 1bfed2828fc5d154d0b96887c28174719bd3864c

File: apps/web/app/routes/app/emergency/new/step1.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+} from "@/components/ui/form";
+import { parseFormData } from "@/utils/read-from-request";
+import { hexSchema } from "@repo/common/schemas";
+import { useState } from "react";
+import { type Hex, padHex } from "viem";
+import { z } from "zod";
+
+export const step1Schema = {
+  title: z.string().min(1, "Cannot be empty"),
+  salt: hexSchema
```
[the remark this answers is not in the document]
Comment (aon) [reply apps/web/app/routes/app/emergency/new/step1.tsx]: Fixed in dc6c55c5de0372b0a19f4a36068da369822b7fd0

[the remark this answers is not in the document]
Comment (aon) [reply apps/web/app/routes/app/emergency/new/step1.tsx]: Fixed in dc6c55c5de0372b0a19f4a36068da369822b7fd0

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (aon) [reply package.json]: Fixed in 63ea8ad497e1fd550b0e81a0bb06ff99beb393bf

File: packages/e2e/package.json [hunk 7]
```diff
@@ -0,0 +1,48 @@
+{
+  "name": "@repo/e2e",
+  "version": "1.0.0",
+  "description": "",
+  "type": "module",
+  "keywords": [],
+  "author": "moonsong-labs",
+  "license": "GPL-3.0-only",
```
[the remark this answers is not in the document]
Comment (aon) [reply packages/e2e/package.json]: Fixed in 63ea8ad497e1fd550b0e81a0bb06ff99beb393bf

File: apps/web/app/.server/service/proposals.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    if (!log.args._proposal) {
+      throw new Error("Invalid log");
+    }
+
+    // If proposal is already stored, we skip it
+    if (storedProposals.some((p) => p.externalId === id)) {
+      continue;
+    }
+
+    const [status, proposalData] = await Promise.all([getUpgradeState(id), getUpgradeStatus(id)]);
```
[the remark this answers is not in the document]
Comment (aon) [reply apps/web/app/.server/service/proposals.ts]: Fixed in 907901797b6784ccf7f942688175d67f0289d8a4

**Comment by aon** [comment]
@anton-matterlabs Please review again the changes. I've made a general refactor for a joint `EthereumConfig` object, per suggestion. Hopefully this explains a little bit more the behaviour inside the app, and clears the difference with the local network. This specific change can be seen in b5975b84e9e60ca5038373613e136ea38b9bcd38

**Review by anton-matterlabs — APPROVED** [review]

**Pull request merged**